lib/datadog/appsec/processor.rb

```diff
@@ -63,6 +87,23 @@ def new_context
         Context.new(self)
       end
 
+      def activate_context
+        existing_active_context = Processor.active_context
+        raise AlreadyActiveContextError if existing_active_context
```

Comment:
This error is never rescued and breaks the app when it raises, is that intended behavior?

Reply:
Hey @idanci. 

Thanks for reporting it. 

Could you elaborate on which scenario this is happening in?

Reply:
Couldn't figure out the root cause. I thought it was because we were using custom instrumentation before the appsec middleware kicked in, but even after removing all the ddtrace/ddstatsd interaction the error is still there.
I don't think the backtrace will be helpful, but here's an example:
```
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/processor.rb:90:in `activate_context'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:35:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/tracing/contrib/rack/middlewares.rb:70:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/sinatra-3.0.5/lib/sinatra/base.rb:2007:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:74:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:58:in `each'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:58:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-parser-0.7.0/lib/rack/parser.rb:24:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-utf8_sanitizer-1.8.0/lib/rack/utf8_sanitizer.rb:28:in `call'",
"/home/wetransfer/frontend/lib/middleware/refuse_non_json_or_multipart_requests.rb:7:in `call'",
"/home/wetransfer/frontend/lib/middleware/json_rescue.rb:38:in `call_with_exception_handling'",
"/home/wetransfer/frontend/lib/middleware/json_rescue.rb:17:in `call'",
"/home/wetransfer/frontend/lib/middleware/x_frame_options_deny.rb:5:in `call'",
"/home/wetransfer/frontend/lib/api/v4.rb:41:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/routing/mapper.rb:20:in `block in <class:Constraints>'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/routing/mapper.rb:49:in `serve'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/journey/router.rb:50:in `block in serve'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/journey/router.rb:32:in `each'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/journey/router.rb:32:in `serve'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/routing/route_set.rb:842:in `call'",
"/home/wetransfer/frontend/config/initializers/redirect_invalid_subdomains.rb:22:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/sanitize_user_agent_header-0.1.0/lib/sanitize_user_agent_header.rb:21:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/deflater.rb:44:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector.rb:53:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector/resolver.rb:71:in `block (2 levels) in write_to_primary'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.1/lib/active_support/notifications/instrumenter.rb:24:in `instrument'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector/resolver.rb:70:in `block in write_to_primary'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/connection_handling.rb:388:in `with_role_and_shard'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/connection_handling.rb:175:in `connected_to'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector/resolver.rb:69:in `write_to_primary'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector/resolver.rb:44:in `write'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector.rb:65:in `select_database'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activerecord-6.1.7.1/lib/active_record/middleware/database_selector.rb:52:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/tempfile_reaper.rb:15:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/etag.rb:27:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/conditional_get.rb:27:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/head.rb:12:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/http/permissions_policy.rb:22:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/http/content_security_policy.rb:19:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/strict_request_uri-1.0.3/lib/strict_request_uri.rb:36:in `call'",
"/home/wetransfer/frontend/config/initializers/set_amplitude_user_properties.rb:7:in `call'",
"/home/wetransfer/frontend/config/initializers/set_amplitude_device_id.rb:7:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/session/abstract/id.rb:266:in `context'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/session/abstract/id.rb:260:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/cookies.rb:697:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.1/lib/active_support/callbacks.rb:98:in `run_callbacks'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/callbacks.rb:26:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'",
"/home/wetransfer/frontend/config/initializers/client_ip_injection.rb:46:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/tracing/contrib/rails/middlewares.rb:17:in `call'",
"/home/wetransfer/frontend/config/initializers/suppress_unacceptable_format.rb:8:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/lograge-0.12.0/lib/lograge/rails_ext/rack/logger.rb:18:in `call_app'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/railties-6.1.7.1/lib/rails/rack/logger.rb:26:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.1/lib/active_support/tagged_logging.rb:99:in `block in tagged'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.1/lib/active_support/tagged_logging.rb:37:in `tagged'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.1/lib/active_support/tagged_logging.rb:99:in `tagged'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/railties-6.1.7.1/lib/rails/rack/logger.rb:26:in `call'",
"/home/wetransfer/frontend/config/initializers/quiet_assets.rb:6:in `call_with_quiet_assets'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/request_store-1.5.1/lib/request_store/middleware.rb:19:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/request_id.rb:26:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/method_override.rb:24:in `call'",
"/home/wetransfer/frontend/config/initializers/host_ident.rb:18:in `call'",
"/home/wetransfer/frontend/config/initializers/i18n.rb:20:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/runtime.rb:22:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/executor.rb:14:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/static.rb:24:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/sendfile.rb:110:in `call'",
"/home/wetransfer/frontend/config/initializers/accept_header_fix.rb:7:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/actionpack-6.1.7.1/lib/action_dispatch/middleware/host_authorization.rb:148:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-cors-2.0.0/lib/rack/cors.rb:102:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:44:in `block (2 levels) in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/instrumentation/gateway.rb:37:in `block in push'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/gateway/watcher.rb:56:in `block in watch_request'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/instrumentation/gateway.rb:19:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/instrumentation/gateway.rb:43:in `block (2 levels) in push'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/instrumentation/gateway.rb:47:in `push'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:43:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:42:in `catch'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:42:in `call'",
"/home/wetransfer/frontend/config/initializers/referer_tracking_middleware.rb:5:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/tracing/contrib/rack/middlewares.rb:98:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/railties-6.1.7.1/lib/rails/engine.rb:539:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:74:in `block in call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:58:in `each'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/rack-2.2.6.4/lib/rack/urlmap.rb:58:in `call'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/unicorn-6.1.0/lib/unicorn/http_server.rb:634:in `process_client'",
"config/unicorn.rb:33:in `block (2 levels) in reload'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/unicorn-6.1.0/lib/unicorn/http_server.rb:739:in `worker_loop'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/unicorn-6.1.0/lib/unicorn/http_server.rb:547:in `spawn_missing_workers'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/unicorn-6.1.0/lib/unicorn/http_server.rb:143:in `start'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/unicorn-6.1.0/bin/unicorn:128:in `<top (required)>'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/bin/unicorn:25:in `load'",
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/bin/unicorn:25:in `<top (required)>'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli/exec.rb:58:in `load'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli/exec.rb:58:in `kernel_load'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli/exec.rb:23:in `run'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli.rb:479:in `exec'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli.rb:31:in `dispatch'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/cli.rb:25:in `start'",
"/usr/local/bundle/gems/bundler-2.2.33/exe/bundle:49:in `block in <top (required)>'",
"/usr/local/bundle/gems/bundler-2.2.33/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'",
"/usr/local/bundle/gems/bundler-2.2.33/exe/bundle:37:in `<top (required)>'",
"/usr/local/bundle/bin/bundle:25:in `load'",
"/usr/local/bundle/bin/bundle:25:in `<main>'"

Please let me know if there is a way to evade appsec errors, it breaks our app. If there is no mitigation, let me know if we need to open a ticket with the dd support instead.

Thanks 🙌 
```

Reply:
@idanci, thanks for sharing the backtrace. 

Could you share more information about your system? 
From the backtrace, I see you are using dd-trace-rb `1.10.1`
Is it a Sinatra app? 

Could you share your `Datadog.configure` block?

Reply:
rails 6.1.7.1
ruby 3.2
ddtrace 1.10.1
dogstatsd-ruby 5.4.0

We use rails and mount several sinatra apps within rails. Those sinatra apps also have home custom middlewares on top of default stack.

```ruby
require 'ddtrace/auto_instrument' unless Rails.env.test?

Datadog.configure do |c|
  options = { service_name: 'frontend' }

  c.service = options[:service_name]
  c.tracing.enabled = true

  if ENV['DD_PROFILING_ENABLED'].present?
    c.profiling.enabled = true
    c.profiling.advanced.force_enable_new_profiler = true
  end

  c.tracing.instrument :sinatra, options.merge(resource_script_names: true)
  c.tracing.instrument :active_support, cache_service: options[:service_name]
  c.tracing.instrument :active_record, options
  c.tracing.instrument :rails, options unless Rails.env.test?
  c.tracing.instrument :faraday, options
  c.tracing.instrument :rest_client, options
  c.tracing.instrument :mysql2, options
  c.tracing.instrument :redis, options
  c.tracing.instrument :active_job, options
  c.tracing.instrument :rake, options.merge(tasks: %w[list of our tasks])

  if Rails.env.production? # we disabled this for now, it breaks the app
    require 'datadog/appsec'

    c.appsec.enabled = true
    c.appsec.instrument :rack
    c.appsec.instrument :rails
    c.appsec.instrument :sinatra
  end
end

class DatadogCustomTraceProcessor
  def call(trace)
    trace.spans.each do |span|
      next unless span.status == Datadog::Tracing::Metadata::Ext::Errors::STATUS

      error_stack = span.get_tag('error.stack').dup
      next if error_stack.blank?

      span.set_tag('error.original_stack', error_stack)
      span.set_tag('error.stack', backtrace_for(error_stack))
    end

    trace
  end

  private

  def backtrace_for(error_stack)
    cleaned_stack = backtrace_cleaner.clean(error_stack.split("\n\tfrom "))
    cleaned_stack.map(&:strip).join("\n")
  end

  def backtrace_cleaner
    @backtrace_cleaner ||= ActiveSupport::BacktraceCleaner.new.tap do |bc|
      bc.add_filter   { |line| line.gsub(Rails.root.to_s, 'frontend') } # strip the Rails.root prefix
      bc.add_silencer { |line| /\/gems\/|vendor\/bundle|usr\/local/.match?(line) } # skip irrelevant lines
    end
  end
end

Datadog::Tracing.before_flush(DatadogCustomTraceProcessor.new)
```

Reply:
@idanci thanks for sharing your configuration. 

That helped me understand what is going on. 

The reason why you see that exception is that when we instrument the Sinatra applications and the rails application, we introduce our AppSec middleware. Since you have multiple apps, there are numerous middleware stacks one per application. 

Our AppSec code expects the request middleware to be called once per request. That allows us to create and dispose of the context safely. 

We can see on your backtrace that the appsec request middleware gets called twice:
```
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:42:in `call'"
```

and 

```
"/home/wetransfer/frontend/vendor/bundle/ruby/3.2.0/gems/ddtrace-1.10.1/lib/datadog/appsec/contrib/rack/request_middleware.rb:35:in `call'"
```

I want to make clear that **we should support your current setup**. Unfortunately, we cannot fix this issue short-term due to some delivery, and the entire team is working on it.  We expect to tackle this issue during Q2

In the meantime, could I suggest removing the `c.appsec.instrument :sinatra` from your configuration? And check if the error continues to occur.

AppSec would continue to run and be executed, and your nested sinatra apps will be protected by the Rack integration and continue reporting any malicious attacks your application would experience.

Reply:
Can confirm that removing `c.appsec.instrument :sinatra` eliminates the error.
What will be the best way to proceed? Should we revert to an older ddtrace version or keep the current one with sinatra appsec disabled and wait for the ddtrace update?

Reply:
We disabled appsec completely due to performance issues, will try to debug it when there's time.
I can open a GH issue, can you please advise what info to put there?
There are no datadog traces for the requests that failed because they were cut off on the client side due to timeouts. We can't spot the issue with ddtrace because ddtrace itself has an issue 😬 . The problem as well might be in our ddtrace config that is shared above, but nothing seems suspicious there.

Thanks for looking into it 🙌

Reply:
@idanci 

> I can open a GH issue, can you please advise what info to put there?

Mentioning your current Datadog configuration, and mentioning what you experience when you upgraded from `1.10.0` to `1.10.1` would be enough. 

We would look into our reliability environment to figure out what change could have caused that degradation between releases.

Thank you so much for your patience 🙌

Reply:
I created a GH issue to track this: https://github.com/DataDog/dd-trace-rb/issues/2722